### DISCUSSION TITLE
ViewState for leafmap.Map

### DISCUSSION THREAD

**Discussion by Rhyzhang** [opengeos/leafmap#150] [Q&A] [answered]
Hi Professor Wu,

I was wondering if it is possible to implement view state on leafmap.Map().

I want to implement this view_state:
`view_state = foo.ViewState(
    longitude=-74.0,
    latitude=40.7,
    zoom=11,
    max_zoom=16,
    pitch=45,
    bearing=0)`

On this code:
`m = leafmap.Map(center=[40, -100], zoom=2, height=600, widescreen=False)
m.to_streamlit(800, 600)`

Sorry, if this was a naive question! I am inexperienced with working with geospatial data.
I understand if you are unable to respond ASAP. I hope you have a Merry Christmas!

Thank you,
Ryan Zhang

**Comment by giswqs** [comment]
`ViewState` is only available for 3D mapping libraries (e.g., pydeck, kepler.gl). folium and ipyleaflet do not have `ViewState`.

**Reply by Rhyzhang** [reply]
Hi Professor Wu,

Merry Christmas and thank you for the quick reply!
I forgot to mention that I am using kepler.gl.
Also, I have figured some things out on my own.

Currently, my code is like this:
```python
import leafmap.kepler as leafmap
m = leafmap.Map(center=[40.7, -74.7], height=600, widescreen=False ,zoom=11)
m.to_streamlit(800, 600)
```

I want it to be something like this:
```python
import leafmap.kepler as leafmap
m = leafmap.Map(center=[40.7, -74.7], height=600, widescreen=False , zoom=11, 
    max_zoom=16,
    pitch=45,
    bearing=0
)
m.to_streamlit(800, 600)
```


However, in kepler.py in leafmap the _config_ for the _Map_ class looks like this:
```python
config = {
    "version": "v1",
    "config": {
        "mapState": {
            "latitude": kwargs["center"][0],
            "longitude": kwargs["center"][1],
            "zoom": kwargs["zoom"],
        }
    },
}
```
_max_zoom_, _pitch_, _bearing_ does not seem to be a kwargs option. Does that mean these config are not possible?
Maybe they can be an added feature?


Ryan

**Comment by giswqs** [comment, marked as answer]
Good suggestion. This feature has been implemented https://github.com/giswqs/leafmap/commit/3524dbabf1ce2b8a4a01c6452c8f352502b18856. Please update the package using `leafmap.update_package()` and restart the kernel to take effect.

**Reply by Rhyzhang** [reply]
Thank you!

May I ask if I can contribute to leafmap in the future if I find any new features that can be improved on?
I love how leafmap is one place for all toolbox for geospatial data and I plan to use it for most of my geospatial projects.

**Reply by giswqs** [reply]
Contributions are very welcome. Please feel free to submit feature requests or pull requests. See https://leafmap.org/contributing
